### PR TITLE
Add WSDS (Warmup-Stable-Decay-Simplified) Scheduler

### PR DESCRIPTION
**Pull request by yanhong-lbh**
### Summary
This PR introduces a new learning rate scheduler `WSDS` designed for continual pretraining scenarios with multiple training periods. The scheduler supports a single warmup phase at the beginning, followed by multiple periods where each period consists of a stable phase and a decay phase. The learning rate resets to the peak value at the start of each new period and decays to exactly 0.0 at period boundaries, making it ideal for checkpointing between training periods.

### Motivation
Continual pretraining often involves multiple distinct training periods (e.g., training on different data distributions or curriculum stages). The WSDS scheduler is designed to:
- Support clean period boundaries where LR reaches exactly 0.0 for optimal checkpointing
- Reset LR to peak at the start of each new period for fresh learning dynamics
- Provide a simple, predictable pattern: warmup (period 0 only) → stable → decay → reset

#### Core Implementation
- **Added `WSDS` scheduler class** in `src/olmo_core/optim/scheduler.py`
  - Supports multiple periods with configurable lengths in steps or tokens
  - Single warmup phase at the beginning (period 0 only)
  - Each period has a stable phase followed by a decay phase
  - LR resets to `initial_lr` at the start of each period
  - LR reaches exactly 0.0 at period boundaries (when `decay_min_lr <= 0`)

#### Configuration Options
- `period_lengths`: List of period lengths in chosen units (steps or tokens)
- `warmup` / `warmup_fraction`: Warmup specification (absolute or fractional, mutually exclusive)
- `decay` / `decay_fraction`: Decay specification per period (absolute or fractional, mutually exclusive)
- `warmup_min_lr`: Starting LR for warmup phase (default: 0.0)
- `decay_min_lr`: Ending LR for decay phase (default: 0.0)

#### Validation
- Validates that warmup + decay does not exceed first period length
- Validates that decay does not exceed any period length
- Clear error messages guide users to fix configuration issues
- No silent adjustments or clamping that could lead to unexpected behavior

#### Testing
- **Added comprehensive test suite** in `src/test/optim/wsds_scheduler_test.py`
  - 24 test cases covering functional behavior, edge cases, and configuration validation
  - Tests for single and multi-period scenarios
  - Tests for both absolute and fractional warmup/decay specifications
  - Parametrized tests for various period counts and lengths
  - Validation tests ensuring all error cases are properly caught

### Example Usage

```python
from olmo_core.optim import WSDS

# Define a scheduler with 3 periods of 1000 steps each
scheduler = WSDS(
    period_lengths=[1000, 1000, 1000],
    warmup=100,              # 100 steps warmup in period 0
    decay_fraction=0.2,      # 20% decay in each period
    warmup_min_lr=0.0,
    decay_min_lr=0.0,        # LR reaches 0 at period boundaries
)

# Get LR at any training step
lr = scheduler.get_lr(initial_lr=1e-3, current=500, t_max=3000)

```

### Experimental Validation

The scheduler has been validated in a real training run:

**Beaker Run:**  
https://beaker.allen.ai/orgs/ai2/workspaces/linear-rnns/work/01K7G7CS4TTP99W2PAFJR4PPD0?taskId=01K7G7CS50E244BJ4BTBS3GRMY&jobId=01K7G7CS8KGB29JG339HSYXHP3

**W&B Report:**  
https://api.wandb.ai/links/yanhong-lbh/hfrqmeyj

**Checkpoints:**  
Verified that checkpoints are created at period boundaries where LR=0:
```bash
/weka/oe-training-default/ai2-llm/checkpoints/yanhongl/linear-rnns/test_v1/
├── step0    # Initial checkpoint
├── step76   # Period 1 boundary (LR=0)
├── step152  # Period 2 boundary (LR=0)
└── step305  # Period 3 boundary (LR=0)
```

The training run confirms that:
- LR properly resets to peak at the start of each period
- LR reaches exactly 0.0 at period boundaries as designed
- Checkpointing at LR=0 boundaries works as expected

### Test Results
- ✅ All 24 WSDS scheduler tests pass
- ✅ All existing optim module tests pass
- ✅ No regressions in other modules

Note: Pre-existing infrastructure test failures (S3/GCS access, distributed GLOO) are unrelated to this PR and are environment-specific issues.

### Changelog
Updated `CHANGELOG.md`:
```markdown
### Added
- Added the `WSDS` (Warmup-Stable-Decay-Simplified) scheduler for continual pretraining with multiple periods.
```

---

> [!NOTE]
> Add `WSDS` multi-period LR scheduler and a `ListCheckpointerCallback` for step-specific checkpointing, with exports, tests, and changelog update.
> 
> - **Optim**:
>   - Added `WSDS` scheduler in `olmo_core/optim/scheduler.py` (warmup → stable → decay across multiple periods; first-period warmup only; configurable absolute/fractional warmup/decay; min LRs).
>   - Exported `WSDS` from `olmo_core/optim/__init__.py`.
> - **Training Callbacks**:
>   - Added `ListCheckpointerCallback` to save checkpoints at exact `save_steps` in `train/callbacks/list_checkpointer.py` and exported in `train/callbacks/__init__.py`.
> - **Tests**:
>   - Extended `src/test/optim/scheduler_test.py` with comprehensive `WSDS` tests (functional behavior, edge cases, parameterization).
> - **Changelog**:
>   - Noted addition of `WSDS` scheduler in `CHANGELOG.md`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 553960cfb8fe80c85e728ff8de5f59619880cfb4. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by viking-sudo-rm** [comment]
Thanks @yanhong-lbh , this looks great! (for more context for @epwalsh, here's the design doc that motivated this: https://docs.google.com/document/d/1pCdq1b6kkw81n3WNY1rq1CHgtotG2hIVdeZPLPLnXYA/edit?usp=sharing)

The main question I have is whether it's straightforward to configure a config to take checkpoints exactly at the bottom of the decay. Is this what `get_cycle_info()` is meant for, and do you have an example from your run?

I'll hold off on approving because there are some failed style checks and @epwalsh knows way more about the codebase than I do.

File: src/olmo_core/optim/scheduler.py [hunk 1]
```diff
@@ -544,3 +544,203 @@ def get_lr(
...
+        self, initial_lr: Union[float, torch.Tensor], current: int, t_max: int
+    ) -> Union[float, torch.Tensor]:
+        """
+        Get learning rate for the current step within repeated WSD cycles.
+        
+        Each cycle of length `cycle_length` follows a WSD pattern, with the cycle
+        ending exactly when LR reaches decay_min_lr (ideal for checkpointing).
+        """
+        # Determine which cycle we're in and position within that cycle
+        cycle_idx = current // self.cycle_length
```
Comment (viking-sudo-rm) [inline src/olmo_core/optim/scheduler.py]: Based on the WSD-S paper, it looks like there is a single warmup applied in advance followed by a repeated high-LR phase, which is different from this.

https://arxiv.org/pdf/2410.05192, Figure 2

File: src/olmo_core/optim/scheduler.py [hunk 2]
```diff
@@ -544,3 +544,203 @@ def get_lr(
...
+    decay_fraction: Optional[float] = 0.1
+    """Fraction of each cycle used for decay, or absolute number of steps/tokens."""
+    
+    warmup_min_lr: float = 0.0
+    """Minimum LR at the start of each warmup phase."""
+    
+    decay_min_lr: float = 0.0
+    """Minimum LR at the end of each decay phase (checkpoint point)."""
+    
+    restart_warmup_from_min: bool = True
```
Comment (viking-sudo-rm) [inline src/olmo_core/optim/scheduler.py]: In https://arxiv.org/pdf/2410.05192 Figure 2, it looks like they restart from *max_lr* rather than warming up again

**Comment by epwalsh** [comment]
Also take a look here for how to fix style/lint/type checks: https://github.com/allenai/OLMo-core/blob/main/CONTRIBUTING.md#formatting-and-linting

File: src/olmo_core/optim/scheduler.py [hunk 3]
```diff
@@ -544,3 +544,181 @@ def get_lr(
...
+                return _linear_decay(initial_lr, D - t, D, self.decay_min_lr)
+        else:
+            D = self._resolve_decay(Li)
+            S = Li - D
+
+            if pos < S:
+                return initial_lr
+            else:
+                t = pos - S
+                return _linear_decay(initial_lr, D - t, D, self.decay_min_lr)
```
[the remark this answers is not in the document]
Comment (viking-sudo-rm) [reply src/olmo_core/optim/scheduler.py]: This doesn't seem like a real issue.

File: src/olmo_core/optim/scheduler.py [hunk 4]
```diff
@@ -544,3 +544,181 @@ def get_lr(
...
+    Note: The peak learning rate is provided via the 'initial_lr' parameter in get_lr().
+      This value remains constant and is reset at the start of each period.
+    """
+
+    period_lengths: List[int] = field(default_factory=list)
+
+    warmup: Optional[int] = None
+    warmup_fraction: Optional[float] = None
+
+    decay: Optional[int] = None
```
Comment (viking-sudo-rm) [inline src/olmo_core/optim/scheduler.py:562]: Related to the divide by 0 thing below, why is this `Optional` and what is the intended behavior when it's not set? No decay?

File: src/olmo_core/train/callbacks/zero_lr_checkpointer.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+from dataclasses import dataclass, field
+from typing import List, Optional, Set
+
+from .checkpointer import CheckpointerCallback
+
+log = logging.getLogger(__name__)
+
+
+@dataclass
+class ZeroLRCheckpointerCallback(CheckpointerCallback):
```
Comment (viking-sudo-rm) [inline src/olmo_core/train/callbacks/zero_lr_checkpointer.py]: Very minor comment but could we rename this to `ListCheckpointerCallback`? We happen to be using it by passing a list where LR is 0, but if I understand correctly, the class is more general than that (it would work for any list)

File: src/olmo_core/optim/scheduler.py (hunk 4 above)
Comment (viking-sudo-rm) [reply src/olmo_core/optim/scheduler.py:562]: Ah, I see. Both are optional because if one is set, the other is ignored?

Comment (yanhong-lbh) [reply src/olmo_core/optim/scheduler.py:562]: Oh! Both are optional because there's a check on line 618 to enforce you cannot set both or set neither, and neither field is required individually

File: src/olmo_core/train/callbacks/zero_lr_checkpointer.py (hunk 5 above)
Comment (yanhong-lbh) [reply src/olmo_core/train/callbacks/zero_lr_checkpointer.py]: Yes!! Just renamed!

**Comment by viking-sudo-rm** [comment]
@epwalsh this LGTM. I'll defer to you to approve in case you have any other concerns

File: src/olmo_core/optim/scheduler.py [hunk 6]
```diff
@@ -544,3 +544,181 @@ def get_lr(
...
+          warmup length W = 0
+          decay  length D = self.decay or round(self.decay_fraction * L_i)
+          stable length   = L_i - D
+
+    Notes:
+      - warmup + decay must not exceed the first period length L_0.
+      - decay must not exceed any period length L_i.
+      - Configuration errors will be raised if these constraints are violated.
+
+    Attributes
```
Comment (tyler-romero) [inline src/olmo_core/optim/scheduler.py]: See elsewhere in the repo for the correct docstring format -  these docstrings are used to autogenerate the documentation site: https://olmo-core.readthedocs.io/en/latest/optim.html#olmo_core.optim.WSD

File: src/olmo_core/optim/scheduler.py [hunk 7]
```diff
@@ -544,3 +544,181 @@ def get_lr(
 
         assert t_max > 0
         return self.schedulers[-1].get_lr(initial_lr, current, t_max)
+
+
+@dataclass
+class WSDS(Scheduler):
```
Comment (tyler-romero) [inline src/olmo_core/optim/scheduler.py:551]: nit: Can we call this something like WSDP (Warmup-Stable-Decay-Periodic) or something more descriptive? It seems like this is a more complicated extension of WSD rather than a simpler one.

Comment (viking-sudo-rm) [reply src/olmo_core/optim/scheduler.py:551]: The name WSD-S comes from this paper: https://arxiv.org/abs/2410.05192

Comment (viking-sudo-rm) [reply src/olmo_core/optim/scheduler.py:551]: (But I suppose we could call it something else and add a note in the comment)

Comment (tyler-romero) [reply src/olmo_core/optim/scheduler.py:551]: Ah thanks for the reference! In that case lets keep the name from the paper

File: src/olmo_core/optim/scheduler.py (hunk 6 above)
Comment (yanhong-lbh) [reply src/olmo_core/optim/scheduler.py]: Thank you! Fixed!

File: src/olmo_core/optim/scheduler.py [hunk 8]
```diff
@@ -544,3 +544,126 @@ def get_lr(
...
+        # decay validation (applies to ALL periods)
+        if (self.decay is None) == (self.decay_fraction is None):
+            raise OLMoConfigurationError(
+                "Exactly one of 'decay' or 'decay_fraction' must be specified."
+            )
+        if self.decay_fraction is not None and not (0.0 <= self.decay_fraction <= 1.0):
+            raise OLMoConfigurationError("'decay_fraction' must be in [0, 1].")
+
+        # Precompute cumulative ends
+        self._cum_period_end = []
```
Comment (2015aroras) [inline src/olmo_core/optim/scheduler.py]: nit: you could instead do something like.
```
self._cum_period_end = np.cumsum(self.period_lengths).tolist()
```

File: src/olmo_core/optim/scheduler.py [hunk 9]
```diff
@@ -544,3 +544,126 @@ def get_lr(
...
+            if x <= end:
+                return idx
+        return len(self._cum_period_end) - 1
+
+    def get_lr(
+        self, initial_lr: Union[float, torch.Tensor], current: int, t_max: int
+    ) -> Union[float, torch.Tensor]:
+        total = self._cum_period_end[-1]
+        if current >= total:
+            return 0.0 if self.decay_min_lr <= 0.0 else self.decay_min_lr
```
[the remark this answers is not in the document]
Comment (2015aroras) [reply src/olmo_core/optim/scheduler.py]: I agree with this. I'd rather you throw an error at post init if `self.decay_min_lr < 0`

File: src/olmo_core/optim/scheduler.py [hunk 10]
```diff
@@ -544,3 +544,126 @@ def get_lr(
...
+
+        pidx = self._find_period(current)
+        start = 0 if pidx == 0 else self._cum_period_end[pidx - 1]
+        Li = self.period_lengths[pidx]
+        pos = min(max(current - start, 0), Li)
+
+        if pidx == 0:
+            W = self._resolve_warmup(Li)
```
Comment (2015aroras) [inline src/olmo_core/optim/scheduler.py]: For consistency with other scheduler implementations, I would recommend treating warmup in a separate if statement and then removing it from the `current` and `t_max`. Then the first period won't need a separate if statement to the other periods (remember to adjust period lengths and `_cum_period_end` accordingly). Something like:

```
if current < warmup:
            return _linear_warmup(initial_lr, current, warmup, self.warmup_min_lr)
elif current >= t_max:
            return self.decay_min_lr
else:
            current = current - warmup
            t_max = t_max - warmup
            // Period logic, which will be the same for L0 and all the other Li
            ...
```

Doing the adjustments will also make your `post_init` validation simpler.

Comment (yanhong-lbh) [reply src/olmo_core/optim/scheduler.py]: done!

File: src/olmo_core/optim/scheduler.py (hunk 8 above)
Comment (yanhong-lbh) [reply src/olmo_core/optim/scheduler.py]: thanks! fixed!

File: src/olmo_core/optim/scheduler.py [hunk 11]
```diff
@@ -544,3 +545,118 @@ def get_lr(
 
         assert t_max > 0
         return self.schedulers[-1].get_lr(initial_lr, current, t_max)
+
+
+@dataclass
+class WSDS(Scheduler):
+    """
+    Warmup–Stable–Decay—Simplified (WSD‑S) scheduler for continual pretraining.
```
Comment (tyler-romero) [inline src/olmo_core/optim/scheduler.py:553]: nit: might be nice to include a link to the paper that proposed this!

File: src/olmo_core/train/callbacks/list_checkpointer.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    _save_steps_set: Set[int] = field(
+        default_factory=set, init=False, repr=False, metadata={"omegaconf_ignore": True}
+    )
+    _last_saved_step: int = field(default=-1, init=False, repr=False)
+
+    def __post_init__(self):
+        super().__post_init__()
+        if not self.save_steps:
+            raise ValueError("'save_steps' must be provided (list of step indices to checkpoint).")
+        # store as a set
```
Comment (tyler-romero) [inline src/olmo_core/train/callbacks/list_checkpointer.py]: nit: redundant comment

File: src/olmo_core/train/callbacks/list_checkpointer.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+from dataclasses import dataclass, field
+from typing import List, Optional, Set
+
+from .checkpointer import CheckpointerCallback
+
+log = logging.getLogger(__name__)
+
+
+@dataclass
+class ListCheckpointerCallback(CheckpointerCallback):
```
Comment (tyler-romero) [inline src/olmo_core/train/callbacks/list_checkpointer.py:11]: This is a pretty strightforward (and useful) change, might be nicer to just implement it directly in CheckpointerCallback and that way we don't need to think about subclasses of CheckpointerCallback when we make modifications there.

File: src/test/optim/wsds_scheduler_test.py
Comment (tyler-romero) [inline src/test/optim/wsds_scheduler_test.py:1]: Can you move all of these tests under `scheduler_test.py`?

We have a specific test structure where all test filenames correspond to the files they are testing.

Eg `olmo_core/optim/scheduler.py` -> `test/optim/scheduler_test.py`

Comment (yanhong-lbh) [reply src/test/optim/wsds_scheduler_test.py:1]: fixed! thank you!

File: src/test/optim/scheduler_test.py [hunk 14]
```diff
@@ -92,3 +96,253 @@ def test_sequential_scheduler():
...
+            decay_min_lr=0.0,
+        )
+        initial_lr = 1.0
+
+        # Stable phase should be 1000 - 50 - 300 = 650 steps
+        # Phases: warmup [0, 50), stable [50, 700), decay [700, 1000]
+        assert scheduler.get_lr(initial_lr, 50, 1000) == initial_lr  # Start of stable
+        assert scheduler.get_lr(initial_lr, 699, 1000) == initial_lr  # End stable
+        assert scheduler.get_lr(initial_lr, 700, 1000) == initial_lr  # Still stable (boundary)
+        lr_decay = scheduler.get_lr(initial_lr, 701, 1000)  # First step of decay
```
[the remark this answers is not in the document]
Comment (yanhong-lbh) [reply src/test/optim/scheduler_test.py:183]: this is the intended behavior
